[3.11] Bug 1693025: Add CSRF to token request endpoint

stlaz opened the pull request:
This is a 3.11 backport of openshift/origin#22767 which adds CSRF validation to OAuth server token requests created from its web console.

@openshift/sig-auth

enj commented on pkg/oauthserver/oauthserver/auth.go:
```diff
@@ -140,7 +140,7 @@ func (c *OAuthServerConfig) WithOAuth(handler http.Handler) (http.Handler, error
 	server.Install(mux, urls.OpenShiftOAuthAPIPrefix)
 
-	tokenRequestEndpoints := tokenrequest.NewEndpoints(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient)
+	tokenRequestEndpoints := tokenrequest.NewTokenRequest(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient, c.ExtraOAuthConfig.OAuthAccessTokenClient)
```
No `OAuthAccessTokenClient`

enj replied:
ffb4fa19c2f81e85e86f866603da103b69913e88 should not have `c.ExtraOAuthConfig.OAuthAccessTokenClient`

enj commented on pkg/oauthserver/server/grant/grant.go:
```diff
@@ -109,18 +108,16 @@ func (l *Grant) Install(mux oauthserver.Mux, paths ...string) {
-	headers.SetStandardHeaders(w)
-
-	user, ok, err := l.auth.AuthenticateRequest(req)
+	authResponse, ok, err := l.auth.AuthenticateRequest(req)
```
This should still be `user, ok, err := l.auth.AuthenticateRequest(req)`

enj commented on pkg/oauthserver/server/headers/oauthbasic.go:
```diff
@@ -0,0 +1,29 @@
+package headers
```
Undo this commit - the handler chain changes are too aggressive of a change to backport.

enj commented on pkg/oauthserver/server/tokenrequest/tokenrequest.go:
```diff
@@ -19,81 +18,54 @@ import (
-	// initLock guards reads and writes to osinOAuthClient when it could still be nil.
-	initLock sync.Mutex
+
+	// to check if we need the logout link for the bootstrap user
```
You should not have this at all.

stlaz replied:
Slipped past me, thanks.

enj commented on pkg/oauthserver/oauthserver/auth.go:
```diff
@@ -140,7 +140,7 @@ func (c *OAuthServerConfig) WithOAuth(handler http.Handler) (http.Handler, error
 	server.Install(mux, urls.OpenShiftOAuthAPIPrefix)
 
-	tokenRequestEndpoints := tokenrequest.NewEndpoints(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient)
+	tokenRequestEndpoints := tokenrequest.NewTokenRequest(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient, c.ExtraOAuthConfig.OAuthAccessTokenClient, c.getCSRF())
```
Remove `OAuthAccessTokenClient`

enj commented on pkg/oauthserver/oauthserver/oauth_apiserver.go:
```diff
@@ -207,20 +204,26 @@ func (c completedOAuthConfig) New(delegationTarget genericapiserver.DelegationTa
+	handler = headers.WithRestoreAuthorizationHeader(handler)
+
+	// this is the normal kube handler chain
+	handler = genericapiserver.DefaultBuildHandlerChain(handler, genericConfig)
```
Undo (see other comment).

enj commented on pkg/oauthserver/server/grant/grant.go:
```diff
@@ -99,28 +97,21 @@ func NewGrant(csrf csrf.CSRF, auth authenticator.Request, render FormRenderer, c
 	switch req.Method {
 	case "GET":
-		l.handleForm(user, w, req)
+		l.handleForm(authResponse, w, req)
```
Undo

enj commented on pkg/oauthserver/server/grant/grant.go:
```diff
@@ -99,28 +97,21 @@ func NewGrant(csrf csrf.CSRF, auth authenticator.Request, render FormRenderer, c
+		l.handleForm(authResponse, w, req)
 	case "POST":
-		l.handleGrant(user, w, req)
+		l.handleGrant(authResponse, w, req)
```
Undo

stlaz commented:
Hopefully addressed all the comments. I tried to modify the CSRF token in a resent POST request and the server failed properly with being unable to verify it.

stlaz commented:
Fixed formatting so that `test/verify` passes...

stlaz commented:
/retest
unit test failed to build coz it got SIGKILL

enj commented on pkg/oauthserver/oauthserver/auth.go:
```diff
@@ -140,7 +140,7 @@ func (c *OAuthServerConfig) WithOAuth(handler http.Handler) (http.Handler, error
 	server.Install(mux, urls.OpenShiftOAuthAPIPrefix)
 
-	tokenRequestEndpoints := tokenrequest.NewTokenRequest(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient, c.ExtraOAuthConfig.OAuthAccessTokenClient)
+	tokenRequestEndpoints := tokenrequest.NewTokenRequest(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient)
```
And then 5dddf472704af722175fb10e6226f3d6345c181e would not be trying to remove it.

stlaz commented:
Patchset cleanup done and the bounce castle is now slowly being inflated for the backport party.

enj commented:
/lgtm

stlaz commented:
/hold
The registry client test fails consistently, needs investigation

stlaz commented:
@openshift/sig-developer-experience you were enabling/disabling the test in master several times, what shall be done here? Note that the backport will go deep in versions.

stlaz commented:
Disabled the failing test, I'll open BZ for the DevExp team once the PR merges.

stlaz commented:
/test unit

enj commented:
/lgtm
/hold cancel

stlaz commented:
BZ tracking the disabled test: https://bugzilla.redhat.com/show_bug.cgi?id=1719156
